### PR TITLE
dm-lwm2m: ZmP v38 changes

### PR DESCRIPTION
**Pull request by mike-scott**
- Correct OpenThread / 802.15.4 check in Kconfig
- Add OpenThread overlay
- Enable GPIO explicitly now as some boards have had many defaults removed

### CODE REVIEW

File: overlay-ot.conf [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Hack: Disable BT on devices which support both for now
```
Comment (mbolivar) [inline overlay-ot.conf:1]: Can we just change the application Kconfig to say:


```
config BT
	default y if !NET_L2_OPENTHREAD
```

?

File: overlay-ot.conf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+CONFIG_BT=n
+
+CONFIG_CPLUSPLUS=y
+CONFIG_FLASH=y
```
Comment (mbolivar) [inline overlay-ot.conf]: Why are some options given in this overlay with the same values they have in prj.conf? That shouldn't be necessary.

For options which need different options when using IPSP vs. OpenThread, does it make sense to handle in the Kconfig using default values to have a single point of truth instead of relying on kconfig to override prj.conf with overlay-ot.conf values? Seems a little nicer to me and may result in a smaller overlay file which is easier to read.

File: overlay-ot.conf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+CONFIG_NET_RX_STACK_SIZE=5120
+
+# Disable conflicting IEEE802154 L2
+CONFIG_NET_L2_IEEE802154=n
+CONFIG_NET_L2_IEEE802154_SHELL=n
+
+CONFIG_NET_L2_OPENTHREAD=y
+
+CONFIG_NET_CONFIG_MY_IPV6_ADDR=""
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="64:ff9b::0527:53ce"
```
Comment (mbolivar) [inline overlay-ot.conf]: What is this? Could use a comment.

File: overlay-ot.conf [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+# Thread by default registers quite a lot addresses.
+CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=6
+CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=8
+
+#CONFIG_OPENTHREAD_DEBUG=y
+#CONFIG_NET_LOG=y
```
Comment (mbolivar) [inline overlay-ot.conf]: Same here, we've already got a debug helpers section in prj.conf which has some of these (like CONFIG_NET_LOG), why duplicate this?

File: overlay-ot.conf (hunk 1 above)
Comment (mike-scott) [reply overlay-ot.conf:1]: This is actually an open item for Nordic to fix. 
See zephyr/boards/arm/nrf52840_pca10056/nrf52840_pca10056_defconfig line 22
CONFIG_BT=y

**Comment by mike-scott** [comment]
Addressed comments (thanks!)

File: overlay-ot.conf (hunk 2 above)
Comment (mike-scott) [reply overlay-ot.conf]: I went through and removed the duplicates.  I was using the same overlay file for testing the upstream LwM2M client.

File: overlay-ot.conf [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+CONFIG_BT=n
+
+# Enable C++
+CONFIG_CPLUSPLUS=y
+
+# Enable flash page layout extension
+CONFIG_FLASH_PAGE_LAYOUT=y
+
+# Disable Proxy
+CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT=n
```
Comment (mbolivar) [inline overlay-ot.conf]: I still think it would make sense to not override values here that are set in prj.conf if it's possible to set them using defaults in the Kconfig file instead. The .conf files will be smaller and have less duplicated information.

For example, this option is default n, so it seems like it could be enabled for BT only with this line in Kconfig:

```
config LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT
	default y if BT
```

Then it could be removed from prj.conf, overlay-modem.conf, and this file.

I've got similar suggestions elsewhere in this file. What do you think?

File: overlay-ot.conf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+# Disable IPv4
+CONFIG_NET_IPV4=n
```
Comment (mbolivar) [inline overlay-ot.conf]: Since this option is default n already, it seems like this line could only be needed for Kinetis K6X, since there's a 'select NET_IPV4' line under FOTA_DEVICE_SOC_SERIES_KINETIS_K6X.

If that's right, can't this line be deleted and just make FOTA_DEVICE_SOC_SERIES_KINETIS_K6X `depends on !NET_L2_OPENTHREAD`? It seems like K6X won't run the OT configuration no matter what, might as well express the dependency in the Kconfig if so.

File: overlay-ot.conf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+CONFIG_NET_CONFIG_NEED_IPV4=n
+CONFIG_NET_CONFIG_MY_IPV4_ADDR=""
+CONFIG_NET_CONFIG_PEER_IPV4_ADDR=""
```
Comment (mbolivar) [inline overlay-ot.conf]: The above three IPV4 lines would be deletable as well.

File: overlay-ot.conf [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+# OpenThread doesn't use NS/NA
+CONFIG_NET_IPV6_NBR_CACHE=n
+CONFIG_NET_IPV6_MLD=n
+
+# OpenThread requires larger stack sizes
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
+CONFIG_NET_TX_STACK_SIZE=2048
+CONFIG_NET_RX_STACK_SIZE=5120
```
Comment (mbolivar) [inline overlay-ot.conf]: Same here, what do you think about

```
config NET_RX_STACK_SIZE
	default 5120 if NET_L2_OPENTHREAD
	default 2048
```

File: overlay-ot.conf [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+# Disable conflicting IEEE802154 L2
+CONFIG_NET_L2_IEEE802154=n
```
Comment (mbolivar) [inline overlay-ot.conf]: This option already depends on `!NET_L2_OPENTHREAD`; are this and the next line really needed? Seems like they could be removed.

File: overlay-ot.conf [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+CONFIG_NET_L2_IEEE802154_SHELL=n
+
+# Enable OpenThread
+CONFIG_NET_L2_OPENTHREAD=y
+
+# OpenThread will set local IP
+CONFIG_NET_CONFIG_MY_IPV6_ADDR=""
```
Comment (mbolivar) [inline overlay-ot.conf]: Why is this needed? The option has no defaults, so it implicitly defaults to the empty string.

File: overlay-ot.conf [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+# Use 4to6 address for mgmt.foundries.io
+CONFIG_NET_CONFIG_PEER_IPV6_ADDR="64:ff9b::c6c7:6c9f"
+
+# Thread by default registers quite a lot addresses.
+CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=6
```
Comment (mbolivar) [inline overlay-ot.conf]: Similarly, for these, what about something like this?

```
config NET_IF_UNICAST_IPV6_ADDR_COUNT
        default 3 if BT
        default 6 if NET_L2_OPENTHREAD

config NET_IF_MCAST_IPV6_ADDR_COUNT
        default 2 if BT
        default 8 if NET_L2_OPENTHREAD
```

File: overlay-ot.conf (hunk 1 above)
Comment (mbolivar) [reply overlay-ot.conf:1]: Got it, thanks.

**Review by mbolivar — COMMENTED** [review]
Thanks for the updates! Some more detailed comments on my idea about making the overlays smaller.

**Comment by mike-scott** [comment]
@mbolivar Thank you for clarifying, I sort of misunderstood your initial discussion.  Let me take a stab at cleaning this up.

**Comment by mike-scott** [comment]
Updated to include Kconfig / .conf cleanup prior to adding overlay-ot.conf

I compared the .configs between old and new for all of our current targets and there weren't any changes.

**Comment by mike-scott** [comment]
I will be adding the ones that apply to Hawkbit/MQTT there as well.

File: Kconfig [hunk 12]
```diff
@@ -26,7 +26,7 @@ if FOTA_DEVICE_SOC_SERIES_KINETIS_K6X
 
 # IPv6 is enabled by default in Zephyr's Kconfig when networking is selected.
 config NET_IPV6
-	default n
+	default n if FOTA_DEVICE_SOC_SERIES_KINETIS_K6X
```
Comment (mike-scott) [inline Kconfig]: Woops.  Just realized this is already in side a FOTA_DEVICE_SOC_SERIES_KINETIS_K6X block

**Review by mbolivar — APPROVED** [review]
Nice and clean! Thanks a lot for making these changes; the ot overlay is super legible now.
